### PR TITLE
[flutter_tools] remove package HTTP and add pub.dev head check tests

### PR DESCRIPTION
**Pull request by jonahwilliams**
We've seen a resurgence of crashes from package:http. Remove the usage and add tests using the FakeHttpClient.

### CODE REVIEW

File: packages/flutter_tools/test/general.shard/devtools_launcher_test.dart [hunk 1]
```diff
@@ -34,13 +35,56 @@ void main() {
...
+        ),
+      ]),
+      processManager: FakeProcessManager.list(<FakeCommand>[]),
+    );
+
+    final DevToolsServerAddress address = await launcher.serve();
+    expect(address, isNull);
+  });
+
+  testWithoutContext('DevtoolsLauncher pings PUB_HOSTED_URL instead of pub.dev for online check', () async {
```
Comment (jmagman) [inline packages/flutter_tools/test/general.shard/devtools_launcher_test.dart:58]: Test if the `PUB_HOSTED_URL` is malformed?

File: packages/flutter_tools/lib/src/devtools_launcher.dart [hunk 2]
```diff
@@ -53,10 +55,17 @@ class DevtoolsServerLauncher extends DevtoolsLauncher {
...
-          await http.head(Uri.https('pub.dev', ''));
+          uri = Uri.https('pub.dev', '');
+        }
+        final io.HttpClientRequest request = await _httpClient.headUrl(uri);
+        final io.HttpClientResponse response = await request.close();
+        await response.drain<void>();
+        if (response.statusCode != io.HttpStatus.ok) {
+          offline = true;
         }
       } on Exception {
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/devtools_launcher.dart:70]: Ahh, good catch

[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/devtools_launcher.dart:72]: Done, needed to modify the fake http client to behave more similarly to the real http client

File: packages/flutter_tools/test/general.shard/devtools_launcher_test.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/test/general.shard/devtools_launcher_test.dart:58]: Done

**Review by jmagman — APPROVED** [review]
LGTM

**Comment by jonahwilliams** [comment]
Ignoring Linux flutter_plugins since that is fixed on ToT

**Comment by justinmc** [comment]
The tree is red on this PR because of a bunch of analyzer failures and a test failure and perf failure.  Not sure if that has anything to do with the contents of this PR, but let me know if anyone has ideas!

**Pull request merged**